man/smsPrint.1
.TH smsPrint 1 "2008 Feb 22" GNU
.SH NAME
smsPrint - Program to print the SMS analysis files.
.SH SYNOPSIS
.B smsPrint
[-\fIoption\fP ...]
.I smsFile
.SH DESCRIPTION
\fISMS\fP is a set of techniques and software implementations for the
analysis, transformation and synthesis of musical sounds based on a
sinusoidal plus residual model. These techniques can be used for
synthesis, processing and coding applications, while some of the
intermediate results might also be applied to other music related
problems, such as sound source separation, musical acoustics, music
perception, or performance analysis. The basic model and
implementation was developed in the PhD thesis by X. Serra in 1989 and
since then many extensions have been proposed at MTG-UPF and by other
researchers.

This program prints on the screen in ASCII code the output of the smsAnal program.

For background information on how to create and synthesize SMS the analysis files, or what to do with 
it other than recreate the original, see the README included with the SMS package
or visit the SMS homepage at:
\fIhttp://www.iua.upf.es/~sms/\fP

For documentation on how to use smsPrint, type "smsPrint --help"

.SH SEE ALSO
smsAnal(1), smsClean(1), smsSynth(1), smsResample(1)


doc/changes.txt
=-=-=-=-=-=-=- libsms change log -=-=-=-=-=-=

=== June 23, 2009 === version 1.1 beta =====================================================

New features:
	- Can now transform sounds using their spectral envelopes. By setting the appropriate
	flags in an SMS_ModifyParams struct and calling sms_modify, it is possible to shape the
	partials of one sound using the spectral envelope of another (or a linear interpolation
	anywhere between their two envelopes), and to transpose sounds while maintaining the
	original spectral envelope. In fact the envelopes do not have to be related to any
	analysed sounds, they could be generated by any means and then applied to the detected
	partials of a given sound, so purely synthetic envelopes could be used as well.
	- The full SMS analysis/synthesis process is now callable from Python, and a Python
	module called SMS.py has been created that will analyze or synthesize a file with a single
	function call. See the demo script python/AnalysisSynthesis.py for an example of using it.
        - new misc. functions: sms_rms, sms_arrayScalarTempered

API changes:
	- All transformations (including transpostion and stochastic component scaling) should
	now be performed using the sms_modify function. The required transformations are specified
	by using the SMS_ModifyParams structure. Examples of using this (from Python) can be found
	in the file python/modify_examples.py.
        - dB conversion is slightly different.  It is normalized to above 0, using a threshhold (default 0.00001)
          to avoid NaN's.  You can set this with sms_setMagThresh
        - all magnitudes are now stored in linear
        - windows are now left un-normalized by default. Call sms_scaleWindow() to normalize it.
Bug fixes:
	- The stochastic gain calculated during residual analysis was being applied twice during
	resynthesis, fixed this.
	- Also changed residual and spectrum energy calculation in sms_residual so that they use 
	the sum of the square of signal samples rather than just the sum of the samples.
	- Fixed a call to memset is sms_synthesize that was writing more data than intended (writing 
	more than the given array size)
	- made some changes to python/setup.py so that the python module builds on OS X (tested
	on Leopard).

Other:
	- Added some unit tests (in Python) to test/Tests.py
        - smsToYaml is gone. smsPrint now naturally prints in yaml format, and you can save it to file.

=== April 9, 2009 === version 1.05e, experimental release =====================================
This is a release heading in the direction of better modularity, especially in the analysis routines.  The ultimate
goal is to be able to interchange each step of the analysis process (such as f0 estiumation, transient extraction,
etc) with custom code, either in C or python.

Many functions received argument changes so they could be easily wrapped in python using SWG.  I tried to
give every array pointer an argument before it that indicates the size of the array.  In some functions, this seemed
too redundant (such as sms_spectrum for example), so they are custom wrapped in the swig interface file.

In doing this, the release is made incompatible with the last because of many name and argument changes. This will
probably continue to happen through the summer.  If projects break or people think things were better the way before,
please email me.

Notable changes:
        - some files have been moved around in an attempt to organize function types (transforms.c. filters.c, tables.c, etc)
        - some more general functions are in sms.c, such as RMS array calculation and frequency conversion (well-tempored)
        - started replacing float with sfloat so that, one day, the library can be optionally compiled in double precision
        - added support for reading of multi-channel audio files - can select what channel you want to analyze 
        - smsAnal now depends much more on library functions (like sms_initAnalParams)
        - tried to fix documentation errors mentioned to me
        - removed some outdated tools (smsToLisp, smsSynthDet - now in smsSynth)
        - moved around synthesis buffers so that they are properly re-allocated if the sizeHop changes
        - using Mersainne Twister random number generator.
        - sms_analyzeFrame() is supposed to be a function for computing sinusoidal analysis without a circular buffer.
          Its not quite at that point yet.
        - SMS_PeakParams structure has parameters just for sinusoidal analysis (it used to need all of SMS_AnalParams)
        - python test scripts have been added for many things (all names test_*.py)

=== Nov. 15, 2008 === version 1.0, first release ============================================

This release represents a revival of the original code from Xavier Serra, which dates back to the early 1990's.
Care has been taken to make sure the functionality of the algorithms important to SMS have not changed,
while still updating function calls to work with modern day platforms.  Future releases will focus more on incorporating
improvements in the SMS model that have since been discovered. Some changes to the original code include:
        - using libsndfile for sound file handling
        - using OOURA fft routines for real-DFT functions (fftw3 as well, but not fully working)
        - optionally can compile to use FFTW3 fft libary, but it is still in a testing stage and fails in pd externals
        - appended the namespaces sms_* and SMS_ to all globally seen functions, structures and defines/enumerations
        - uses scons instead of make, as it is very powerful, being based in python
        - python SWIG wrappers
        - synthesis parameters "stocgain" and "ftranspose" are added to change the current synthesis, but not the data itself
        - smsToYaml - converts a binary *.sms file to YAML text format, which is very easy to read

tools/smsPrint.c
/* 
 * Copyright (c) 2008 MUSIC TECHNOLOGY GROUP (MTG)
 *                         UNIVERSITAT POMPEU FABRA 
 * 
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of 
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 */

#include "sms.h"
#include <popt.h>

const char *help_header_text =
        "\n\n"
        "Prints an SMS analysis file made with libsms, specified as the last argument.\n"
        "If -o FILENAME or --output FILENAME is provided, the output will be written to a text file.\n"
        "The output is printed according to the YAML specification.\n"
        "\n [OPTIONS] \n";

enum  PRINT_TYPE
{
        PRINT_ALL,
        PRINT_DET,
        PRINT_STOC,
        PRINT_ENV,
        PRINT_HDR,
};

int main (int argc, const char *argv[])
{
	int optc;   /* switch */
        const char *pInName = NULL, *pOutName = NULL;
        SMS_Header *pSmsHeader = NULL;
        FILE *pSmsFile, *fp;
        SMS_Data smsData;
        int iError, i, j, iFormat = PRINT_ALL, iFirstFrame = 0, iLastFrame = -1, 
                iFirstTraj = 0, iLastTraj = -1, inDB = 0;
        float fInitialTime = 0, fEndTime = 0, mag;
	poptContext pc;

        struct poptOption	options[] =
                {
                        {"type", 't', POPT_ARG_INT, &iFormat, 0, 
                         "print type: 0=all (default), 1=deterministic, 2=residual, 3=envelope, 4=header", "int"},
                        {"initial-time", 'i', POPT_ARG_FLOAT, &fInitialTime, 0, 
                         "initial time", "float"},
                        {"end-time", 'e', POPT_ARG_FLOAT, &fEndTime, 0, 
                         "end time", "float"},
                        {"decibels", 'd', POPT_ARG_NONE, &inDB, 0, 
                         "make magnitude values printed i decibels", 0},
                        {"output", 'o', POPT_ARG_STRING, &pOutName, 0, 
                         "output text file name", "FILENAME"},
                        POPT_AUTOHELP
                        POPT_TABLEEND
                };

        pc = poptGetContext("smsPrint", argc, argv, options, 0);
        poptSetOtherOptionHelp(pc, help_header_text);

        if (argc <= 1)
        {
                poptPrintUsage(pc,stderr,0);
                return 1;
        }

        while ((optc = poptGetNextOpt(pc)) > 0) {
/*                 switch (optc) { */
/*                         /\* specific arguments are handled here *\/ */
/*                 } */
        }
        if (optc < -1) 
        {
                /* an error occurred during option processing */
                printf("%s: %s\n",
                       poptBadOption(pc, POPT_BADOPTION_NOALIAS),
                       poptStrerror(optc));
                return 1;
        }
        pInName = poptGetArg(pc);

        if(pOutName == NULL)
                fp = (FILE *) stdout;
        else
        {
                fp = fopen(pOutName, "w");
                printf("writing %s.\n", pOutName);
        }

        sms_getHeader ((char *) pInName, &pSmsHeader, &pSmsFile);
        if(sms_errorCheck())
        {
                printf("error in sms_getHeader: %s \n", sms_errorString());
                printf("failed when trying to open file %s \n", pInName );
                exit(EXIT_FAILURE);
        }	    
        sms_allocFrameH (pSmsHeader, &smsData);
        if(sms_errorCheck())
        {
                printf("error in sms_allocFrameH: %s \n", sms_errorString());
                exit(EXIT_FAILURE);
        }	    
	
        //:::::::::: write Header ::::::::::::::::
        fprintf(fp,"smsHeader:\n");
        fprintf(fp,"    nFrames         : %d\n", pSmsHeader->nFrames);
        fprintf(fp,"    iFrameRate       : %d\n", pSmsHeader->iFrameRate);
        fprintf(fp,"    nTracks    : %d\n", pSmsHeader->nTracks);
        fprintf(fp,"    nStochasticCoeff : %d\n", pSmsHeader->nStochasticCoeff);
        fprintf(fp,"    iFormat          : ");
        if(pSmsHeader->iFormat == SMS_FORMAT_H) 
                fprintf(fp,"harmonic\n");
        else if(pSmsHeader->iFormat == SMS_FORMAT_IH) 
                fprintf(fp,"inharmonic\n");
        else if(pSmsHeader->iFormat == SMS_FORMAT_HP) 
                fprintf(fp,"harmonic_with_phase\n");
        else if(pSmsHeader->iFormat == SMS_FORMAT_IHP) 
                fprintf(fp,"inharmonic_with_phase\n");
        fprintf(fp,"    iStochasticType  : ");
        if(pSmsHeader->iStochasticType == SMS_STOC_IFFT) 
                fprintf(fp,"stft\n");
        else if(pSmsHeader->iStochasticType == SMS_STOC_APPROX) 
                fprintf(fp,"approx\n");
        else if(pSmsHeader->iStochasticType == SMS_STOC_NONE) 
                fprintf(fp,"none\n");
        fprintf(fp,"    nEnvCoeff : %d\n", pSmsHeader->nEnvCoeff);
        fprintf(fp,"    iMaxFreq : %d\n", pSmsHeader->iMaxFreq);
        fprintf(fp,"    iEnvType  : ");
        if(pSmsHeader->iEnvType == SMS_ENV_CEP) 
                fprintf(fp,"cepstrum\n");
        else if(pSmsHeader->iEnvType == SMS_ENV_FBINS) 
                fprintf(fp,"fbins\n");
        else if(pSmsHeader->iEnvType == SMS_ENV_NONE) 
                fprintf(fp,"none\n");
        fprintf(fp,"    iSamplingRate   : %d\n", pSmsHeader->iSamplingRate);  

        //:::::::::::: Write Analysis Arguments :::::::::::::::
        if (pSmsHeader->nTextCharacters > 0)
        {
                char cOneLine[100];
                int t, iStrStart, nBytes;
                fprintf(fp,"\ntext_string: >\n");
                iStrStart = 0;
                for(t = 0; t < pSmsHeader->nTextCharacters; t++)
                {
                        if(pSmsHeader->pChTextCharacters[t] == ',' && (t - iStrStart) > 60)
                        {   
                                memset(cOneLine, 0, 100);
                                nBytes = t + 1 - iStrStart;
                                strncpy (cOneLine, pSmsHeader->pChTextCharacters+ iStrStart, nBytes);
                                fprintf(fp,"    %s\n", cOneLine);
                                iStrStart = t+2;
                        }
                        else if(pSmsHeader->nTextCharacters == t - 1)
                        {
                                memset(cOneLine, 0, 100);
                                nBytes = t + 1 - iStrStart;
                                strncpy (cOneLine, pSmsHeader->pChTextCharacters+ iStrStart, nBytes);
                                fprintf(fp,"    %s\n", cOneLine);
                        }
                }
        } 

        iFirstFrame = 
                MIN (pSmsHeader->nFrames - 1, fInitialTime * pSmsHeader->iFrameRate);
        if (fEndTime > 0) 
                iLastFrame = 
                        MIN (fEndTime * pSmsHeader->iFrameRate, pSmsHeader->nFrames);
        else
                iLastFrame = pSmsHeader->nFrames; 

        if (iFirstTraj > 0)
                iFirstTraj = MIN (pSmsHeader->nTracks, iFirstTraj);

        if (iLastTraj >= 0)
                iLastTraj = MIN (pSmsHeader->nTracks, iLastTraj);
        else
                iLastTraj = pSmsHeader->nTracks;
        if(iFormat != PRINT_HDR)
        {
                //:::::::::::: Write Data :::::::::::::::
                fprintf(fp, "\nsmsData:\n");
                for(i = iFirstFrame; i < iLastFrame; i++)
                {
                        sms_getFrame (pSmsFile, pSmsHeader, i, &smsData);
                        if(sms_errorCheck())
                        {
                                printf("error in sms_getFrame: %s \n", sms_errorString());
                                exit(EXIT_FAILURE);
                        }	    
                        fprintf(fp,"\n  - frame    : %d \n", i);
                        fprintf(fp,"    timetag  : %f \n", (float) i / pSmsHeader->iFrameRate);
                        if(iFormat == PRINT_DET || iFormat == PRINT_ALL)
                        {
                                if(pSmsHeader->iFormat == SMS_FORMAT_H ||
                                   pSmsHeader->iFormat == SMS_FORMAT_HP)
                                        fprintf(fp,"    harmonics:\n");
                                if(pSmsHeader->iFormat == SMS_FORMAT_IH ||
                                   pSmsHeader->iFormat == SMS_FORMAT_IHP)
                                        fprintf(fp,"    tracks:\n");
                                if (pSmsHeader->iFormat == SMS_FORMAT_H ||
                                    pSmsHeader->iFormat == SMS_FORMAT_IH)
                                {	
                                        for(j = iFirstTraj; j < iLastTraj; j++)
                                        {
                                                if(smsData.pFSinAmp[j] > 0.00000001 )
                                                        fprintf(fp, "       %-4d: [%12f, %12f]  \n", j,
                                                                smsData.pFSinFreq[j], smsData.pFSinAmp[j]);
                                        }
                                }
                                else
                                {
                                        for(j = iFirstTraj; j < iLastTraj; j++)
                                        {
                                                if(smsData.pFSinAmp[j] > 0.00000001 )
                                                        fprintf(fp,"        %-4d: [%12f, %12f, %12f]  \n", j,
                                                                smsData.pFSinFreq[j], smsData.pFSinAmp[j], smsData.pFSinPha[j]);	
                                        }
                                }
                        }
		
                        if (iFormat == PRINT_STOC || iFormat == PRINT_ALL) /* \todo will this crash if no stoc component? */
                        {	
                                if( pSmsHeader->iStochasticType == SMS_STOC_APPROX )
                                {
                                        fprintf(fp,"    stocGain: %f\n", *(smsData.pFStocGain));
                                        fprintf(fp,"    stocCoefficients: [");
                                        for(j = 0; j < smsData.nCoeff; j++)
                                        {
                                                if(j%4 == 0 && j !=0)
                                                        fprintf(fp,",\n                       ");
                                                else if( j !=0)
                                                        fprintf(fp,", ");
                                                fprintf(fp,"%9f", smsData.pFStocCoeff[j]);
                                        }
                                        fprintf(fp," ]\n");
                                }   
                                else if(pSmsHeader->iStochasticType == SMS_STOC_IFFT)
                                {
                                        fprintf(fp,"    stocGain: %f\n", *(smsData.pFStocGain));
                                        fprintf(fp,"    residual: [");
                                        for(j = 0; j < smsData.nCoeff; j++)
                                        {
                                                if(j%4 == 0 && j !=0)
                                                        fprintf(fp,",\n                       ");
                                                else if( j !=0)
                                                        fprintf(fp,", ");
                                                fprintf(fp,"[%9f, %9f]", smsData.pFStocCoeff[j], smsData.pResPhase[j]);
                                        }
                                        fprintf(fp," ]\n");
                                }
                        }
                        if (iFormat == PRINT_ALL || iFormat == PRINT_ENV)
                        {
                                if (pSmsHeader->iEnvType == SMS_ENV_CEP)
                                {
                                        fprintf(fp,"    cepstrumCoefficients: [");
                                        for(j = 0; j < smsData.nEnvCoeff; j++)
                                        {
                                                if(j%4 == 0 && j !=0)
                                                        fprintf(fp,",\n                       ");
                                                else if( j !=0)
                                                        fprintf(fp,", ");
                                                fprintf(fp,"%9f", smsData.pSpecEnv[j]);
                                        }
                                        fprintf(fp," ]\n");
                                }
                                if (pSmsHeader->iEnvType == SMS_ENV_FBINS)
                                {
                                        fprintf(fp,"    specEnv: [");
                                        for(j = 0; j < smsData.nEnvCoeff; j++)
                                        {
                                                if(inDB) 
                                                        mag = sms_magToDB(smsData.pSpecEnv[j]);
                                                else 
                                                        mag = smsData.pSpecEnv[j];
                                                if(j%4 == 0 && j !=0)
                                                        fprintf(fp,",\n                       ");
                                                else if( j !=0)
                                                        fprintf(fp,", ");
                                                fprintf(fp,"%9f", mag);
                                        }
                                        fprintf(fp," ]\n");
                                }
                        }
                }
        }
        free (pSmsHeader);
        fclose (fp);
        fclose (pSmsFile);
        return(1);
}
